app/Main.hs

```diff
@@ -86,5 +86,41 @@ testShadowing =
     abs_twice_shadow = Abs "x" $ Abs "y" $ Abs "y" $ Abs "x" $ Term "x"
     app_twice_shadow = App (App (App (App abs_twice_shadow $ Lit "1") $ Lit "2") $ Lit "3") $ Lit "4"
 
+etaReduce :: Expr -> Expr
+etaReduce expr = case expr of
+  (Abs ident (App expr' (Term ident'))) ->
+    if ident == ident' && isNotFreeVar then expr' else expr
+      where isNotFreeVar = not $ isFreeVarOf ident expr'
+  _                                     -> expr
+
+testEtaReduce :: IO ()
+testEtaReduce =
+  print
+    [ etaReduce (Abs "x" (App l_id (Term "x"))) == l_id
+    , etaReduce (Abs "x" (App (App l_const (Term "z")) (Term "x"))) == App l_const (Term "z")
+    , etaReduce not_reducable_1 == not_reducable_1
+    , etaReduce not_reducable_2 == not_reducable_2
+    , etaReduce not_reducable_3 == not_reducable_3
+    ]
+  where
+    not_reducable_1 = Abs "x" (App (Abs "y" (Term "x")) (Term "x"))
+    not_reducable_2 = Abs "x" (App (Abs "y" (Abs "z" (Term "x"))) (Term "x"))
+    not_reducable_3 = Abs "x" (App (App l_const (Term "x")) (Term "x"))
+
+isFreeVarOf :: Identifier -> Expr -> Bool
+isFreeVarOf var (Lit s) = False
```

Comment:
why `False`?

Reply:
I was thinking that semantically its not correct to say that variable is free in some literal, cause literal obviously couldn't contain that variable at all :)

But after your comment I realized that in the context of eta-reduction this case should probably return `True`, so we don't do anything with such incorrect (will `eval` into `Nothing`) abstraction